src/Common/Core/Impl/OS/Win32Process.cs

```diff
@@ -8,11 +8,54 @@
 using System.Security.AccessControl;
 using System.Security.Principal;
 using Microsoft.Win32.SafeHandles;
+using System.Threading;
 using static Microsoft.Common.Core.NativeMethods;
 
 namespace Microsoft.Common.Core.OS {
     public class Win32Process {
-        public static int StartProcessAsUser(WindowsIdentity winIdentity, string applicationName, string commandLine, string workingDirectory, Win32NativeEnvironmentBlock environment, out Stream stdin, out Stream stdout, out Stream stderror) {
+        private Win32Process(PROCESS_INFORMATION pi) {
+            _hasExited = false;
+            ProcessId = pi.dwProcessId;
+            MainThreadId = pi.dwThreadId;
+            _processHandle = new SafeProcessHandle(pi.hProcess, true);
+            _threadHandle = new SafeThreadHandle(pi.hThread);
+            ProcessWaitHandle wait = new ProcessWaitHandle(_processHandle);
+            ThreadPool.RegisterWaitForSingleObject(wait, (o, t) => {
```

Comment:
Also, it looks like they recommend calling `RegisteredWaitHandle.Unregister` in the callback.